operators/src/concurrency/thread_pool.rs

```diff
@@ -0,0 +1,339 @@
+use std::sync::Arc;
+use std::thread::JoinHandle;
+use std::{iter, mem};
+
+use crossbeam::atomic::AtomicCell;
+use crossbeam::deque::{Injector, Steal, Stealer, Worker};
+use crossbeam::sync::WaitGroup;
+use crossbeam::utils::Backoff;
+use failure::_core::marker::PhantomData;
+
+/// A chunk of work with some metadata
+struct Task {
+    _group_id: TaskGroupId,
+    task_fn: Box<dyn TaskFn>,
+}
+
+type TaskGroupId = usize;
+
+pub trait TaskFn: FnOnce() + Send {}
+pub trait StaticTaskFn: TaskFn + 'static {}
+
+impl<T> TaskFn for T where T: FnOnce() + Send {}
+impl<T> StaticTaskFn for T where T: TaskFn + 'static {}
+
+impl Task {
+    /// Create a new task to be executed at some point
+    fn new<F>(group_id: TaskGroupId, f: F) -> Self
+    where
+        F: StaticTaskFn,
+    {
+        Self {
+            _group_id: group_id,
+            task_fn: Box::new(f),
+        }
+    }
+
+    /// Executes the task
+    /// TODO: use `FnTraits` once stable
+    fn call_once(self) {
+        (self.task_fn)()
+    }
+}
+
+/// A worker thread pool for compute-heavy tasks
+///
+/// TODO: increase threads to a certain maximum number if current threads don't produce results fast enough?
+/// TODO: schedule tasks based on group context?
+///
+#[derive(Debug)]
+pub struct ThreadPool {
+    degree_of_parallelism: usize,
+    global_queue: Arc<Injector<Task>>,
+    stealers: Vec<Stealer<Task>>,
+    threads: Vec<JoinHandle<()>>,
+    next_group_id: AtomicCell<usize>,
+}
+
+impl ThreadPool {
```

Comment:
wouldn't it make sense to make this a trait? Then we could have different schedulers depending on workload and software/hardware stack to be configurable.

Reply:
Sounds like premature overengineering, but I could do it if you really want it(?).